File: sdks/ios/Sources/VistreaRuntimeModels/RuntimeIdentifierFactory.swift
import CryptoKit
import Foundation

/// Generates canonical typed UUIDv7 identifiers for runtime-produced models.
public enum RuntimeIdentifierFactory {
    public static func make(prefix: String, date: Date = Date()) -> String {
        var bytes = [UInt8](repeating: 0, count: 16)
        let milliseconds = UInt64(max(0, date.timeIntervalSince1970 * 1_000))
        for index in 0..<6 {
            bytes[5 - index] = UInt8((milliseconds >> UInt64(index * 8)) & 0xff)
        }
        for index in 6..<16 {
            bytes[index] = UInt8.random(in: .min ... .max)
        }
        bytes[6] = (bytes[6] & 0x0f) | 0x70
        bytes[8] = (bytes[8] & 0x3f) | 0x80
        return "\(prefix)_\(format(bytes))"
    }

    public static func deterministic(prefix: String, seed: String) -> String {
        var bytes = Array(SHA256.hash(data: Data(seed.utf8)).prefix(16))
        bytes[6] = (bytes[6] & 0x0f) | 0x70
        bytes[8] = (bytes[8] & 0x3f) | 0x80
        return "\(prefix)_\(format(bytes))"
    }

    private static func format(_ bytes: [UInt8]) -> String {
        let hex = bytes.map { String(format: "%02x", $0) }.joined()
        return [
            String(hex.prefix(8)),
            String(hex.dropFirst(8).prefix(4)),
            String(hex.dropFirst(12).prefix(4)),
            String(hex.dropFirst(16).prefix(4)),
            String(hex.dropFirst(20).prefix(12)),
        ].joined(separator: "-")
    }
}

File: sdks/android/src/main/kotlin/dev/vistrea/protocol/v1/RuntimeIdentifierFactory.kt
package dev.vistrea.protocol.v1

import java.security.MessageDigest
import java.security.SecureRandom
import java.time.Instant

private const val UUID_BYTE_COUNT = 16
private const val UUID_TIMESTAMP_BYTE_COUNT = 6
private const val UUID_VERSION_BYTE_INDEX = 6
private const val UUID_VARIANT_BYTE_INDEX = 8
private const val VERSION_MASK = 0x0f
private const val VERSION_SEVEN = 0x70
private const val VARIANT_MASK = 0x3f
private const val VARIANT_RFC_4122 = 0x80
private const val BYTE_MASK = 0xff
private const val UUID_GROUP_ONE_END = 8
private const val UUID_GROUP_TWO_END = 12
private const val UUID_GROUP_THREE_END = 16
private const val UUID_GROUP_FOUR_END = 20
private const val UUID_HEX_LENGTH = 32

/** Generates canonical typed UUIDv7 identifiers for runtime-produced models. */
object RuntimeIdentifierFactory {
    private val random = SecureRandom()

    fun make(prefix: String, time: Instant = Instant.now()): String {
        val bytes = ByteArray(UUID_BYTE_COUNT)
        val milliseconds = time.toEpochMilli().coerceAtLeast(0)
        for (index in 0 until UUID_TIMESTAMP_BYTE_COUNT) {
            bytes[UUID_TIMESTAMP_BYTE_COUNT - 1 - index] =
                ((milliseconds shr (index * Byte.SIZE_BITS)) and BYTE_MASK.toLong()).toByte()
        }
        val suffix = ByteArray(UUID_BYTE_COUNT - UUID_TIMESTAMP_BYTE_COUNT)
        random.nextBytes(suffix)
        suffix.copyInto(bytes, destinationOffset = UUID_TIMESTAMP_BYTE_COUNT)
        setVersionAndVariant(bytes)
        return "${prefix}_${format(bytes)}"
    }

    fun deterministic(prefix: String, seed: String): String {
        val bytes = MessageDigest.getInstance("SHA-256")
            .digest(seed.toByteArray(Charsets.UTF_8))
            .copyOf(UUID_BYTE_COUNT)
        setVersionAndVariant(bytes)
        return "${prefix}_${format(bytes)}"
    }

    private fun setVersionAndVariant(bytes: ByteArray) {
        bytes[UUID_VERSION_BYTE_INDEX] =
            ((bytes[UUID_VERSION_BYTE_INDEX].toInt() and VERSION_MASK) or VERSION_SEVEN).toByte()
        bytes[UUID_VARIANT_BYTE_INDEX] =
            ((bytes[UUID_VARIANT_BYTE_INDEX].toInt() and VARIANT_MASK) or VARIANT_RFC_4122).toByte()
    }

    private fun format(bytes: ByteArray): String {
        val hex = bytes.joinToString(separator = "") { byte ->
            "%02x".format(byte.toInt() and BYTE_MASK)
        }
        return listOf(
            hex.substring(0, UUID_GROUP_ONE_END),
            hex.substring(UUID_GROUP_ONE_END, UUID_GROUP_TWO_END),
            hex.substring(UUID_GROUP_TWO_END, UUID_GROUP_THREE_END),
            hex.substring(UUID_GROUP_THREE_END, UUID_GROUP_FOUR_END),
            hex.substring(UUID_GROUP_FOUR_END, UUID_HEX_LENGTH),
        ).joinToString("-")
    }
}
